File: macapp/Sources/Reflip/Catalogue.swift
import Foundation

/// What `reflip models` answers in its three modes.
///
/// Nothing worth saying about a model is decided here. The catalogue's two sentences per
/// entry, whether a search result must be refused, and the verdict after a measurement
/// are all reflip's words, read exactly as printed. The one rule this file follows is the
/// same one `Receipt` follows: a figure that was not measured stays absent rather than
/// becoming a zero, because a zero here would be a claim nobody checked.

// MARK: - what `reflip models --recommended --json` answers

/// One entry in the catalogue: a model worth trying, with reflip's own sentences about
/// it.
///
/// This struct is not the list of models this window knows about. It is the shape of one
/// row in whatever list `reflip models --recommended` prints, which is the whole point:
/// the catalogue lives in the Python package, moves on its own schedule, and a build of
/// this window from last month already reads a catalogue entry it had never seen before
/// without being told what one looks like in advance, because the shape does not change,
/// only the rows do.
struct CatalogueModel: Decodable, Equatable, Identifiable {
    let ref: String
    let params: String
    let sizeGB: Double
    let goodAt: String
    let watchOut: String
    let languages: String
    /// What a real measurement on some machine said, in a sentence, or null for a model
    /// nobody has measured yet. Never a number on its own: a bare z-score with no context
    /// is not something a first-time reader can judge.
    let measured: String?
    /// True disqualifies rather than ranks: a model that watermarks its own output is
    /// never actually offered by the catalogue today, but this window does not assume
    /// that stays true and checks the field anyway.
    let watermarks: Bool
    let source: String
    let tags: [String]
    let installed: Bool

    var id: String { ref }

    enum CodingKeys: String, CodingKey {
        case ref, params, languages, measured, watermarks, source, tags, installed
        case sizeGB = "size_gb"
        case goodAt = "good_at"
        case watchOut = "watch_out"
    }
}

/// The catalogue plus what this Mac already has.
struct RecommendedModels: Decodable, Equatable {
    let recommended: [CatalogueModel]
    let defaultModel: String
    let installed: [String]
    /// Why the server itself could not be read, when there is a reason. Not shown as an
    /// error on its own: the catalogue is still worth showing even when the server is
    /// down, which is exactly when somebody most needs to pick a model to download.
    let serverReason: String?

    enum CodingKeys: String, CodingKey {
        case recommended, installed
        case defaultModel = "default"
        case serverReason = "server_reason"
    }

    // Defensive the way `ServerInfo` is, for the same reason: a future reflip that
    // drops or renames a field here should show an empty section rather than fail to
    // decode the whole window.
    init(from decoder: Decoder) throws {
        let box = try decoder.container(keyedBy: CodingKeys.self)
        recommended = try box.decodeIfPresent([CatalogueModel].self, forKey: .recommended) ?? []
        defaultModel = try box.decodeIfPresent(String.self, forKey: .defaultModel) ?? ""
        installed = try box.decodeIfPresent([String].self, forKey: .installed) ?? []
        serverReason = try box.decodeIfPresent(String.self, forKey: .serverReason)
    }
}

// MARK: - what `reflip models --search QUERY --json` answers

struct SearchResult: Decodable, Equatable, Identifiable {
    let ref: String
    let repo: String
    let downloads: Int
    let likes: Int
    let gated: Bool
    let page: String
    /// A sentence, when this model must never be used as the rewriter because it
    /// watermarks its own output, or null. A result carrying one is shown with no
    /// Download button at all: offering the button and refusing the click is a worse
    /// answer than not offering it.
    let refused: String?

    var id: String { ref }
}

struct SearchResults: Decodable, Equatable {
    let query: String
    let results: [SearchResult]
    /// reflip's own caution about the results, when it has one: that these are search
    /// results and not recommendations, or that Hugging Face could not be reached.
    let note: String?

    enum CodingKeys: String, CodingKey {
        case query, results, note
    }

    init(from decoder: Decoder) throws {
        let box = try decoder.container(keyedBy: CodingKeys.self)
        query = try box.decodeIfPresent(String.self, forKey: .query) ?? ""
        results = try box.decodeIfPresent([SearchResult].self, forKey: .results) ?? []
        note = try box.decodeIfPresent(String.self, forKey: .note)
    }
}

// MARK: - what `reflip models --measure MODEL --json` answers

/// Two shapes share one document. A refusal, because the model watermarks its own
/// output or because there is no benchmark corpus on this machine, carries only `ok`,
/// `model` and `reason`; a real measurement carries every figure below and no `reason`.
/// Every numeric field stays optional rather than defaulting to zero for the same
/// reason `Receipt.coverage` does: a zero here is itself a result worth reading, not a
/// stand-in for "this was not decoded".
struct MeasureResult: Decodable, Equatable {
    let ok: Bool
    let model: String
    let reason: String?
    let samples: Int?
    let errors: Int?
    let zBefore: Double?
    let zAfter: Double?
    let coverage: Double?
    let editRatio: Double?
    let seconds: Double?
    let tokensPer1kWords: Double?
    let coverageNote: String?
    let verdict: String?

    enum CodingKeys: String, CodingKey {
        case ok, model, reason, samples, errors, coverage, seconds, verdict
        case zBefore = "z_before"
        case zAfter = "z_after"
        case editRatio = "edit_ratio"
        case tokensPer1kWords = "tokens_per_1k_words"
        case coverageNote = "coverage_note"
    }

    /// A result this window made up because reflip could not be asked at all, or
    /// answered with something that would not decode: the same last resort
    /// `Rewriter.failure` reaches for, in the same shape a real refusal would have used,
    /// so the row that shows it does not need a second code path for "reflip never
    /// answered" versus "reflip answered no".
    static func launchFailure(model: String, reason: String) -> MeasureResult {
        MeasureResult(ok: false, model: model, reason: reason, samples: nil, errors: nil,
                      zBefore: nil, zAfter: nil, coverage: nil, editRatio: nil,
                      seconds: nil, tokensPer1kWords: nil, coverageNote: nil, verdict: nil)
    }
}

File: reflip/catalogue.py
"""Which model to rewrite with, and how to find out rather than be told.

Two things live here. A short list of models worth trying, each with a plain sentence
about what it is good and bad at, and a search over Hugging Face for anything else in a
format the local server can run. Neither of them is a recommendation you have to take on
faith: `reflip models --measure NAME` runs that model over watermarked texts from the
benchmark corpus and reports what the detector said, so the list can be checked and
argued with.

The one hard rule in here is the `watermarks` field. A rewriter that marks its own output
turns the whole exercise into a swap of one watermark for another, so any model that does
is refused rather than ranked.
"""
from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import asdict, dataclass, field

HF_API = "https://huggingface.co/api/models"
HF_TIMEOUT = 15.0


@dataclass
class Model:
    """One candidate rewriter, described the way a person would ask about it."""

    ref: str                      # what to pass to `reflip pull`
    params: str                   # "4B", "8B", a mixture of experts
    size_gb: float                # the download, at the quantisation named in ref
    good_at: str                  # one sentence
    watch_out: str                # one sentence, always present: nothing here is perfect
    languages: str
    measured: str | None = None   # what a measurement on this machine said, or None
    watermarks: bool = False      # true means it is disqualified, not ranked
    source: str = "ollama"        # "ollama" library, or "hf" for a Hugging Face repository
    tags: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return asdict(self)


# Ordered by how well they fit the job: a rewriter needs to be fluent and fast, and
# needs no reasoning at all. A thinking model spends its tokens deciding how to
# paraphrase, which is why the reasoning variants are further down than their size
# suggests. "measured" is filled in only where a number was actually produced on this
# machine and can be reproduced with `reflip models --measure`.
CATALOGUE: list[Model] = [
    Model(
        ref="qwen3:4b-instruct-2507-q4_K_M", params="4B", size_gb=2.5,
        good_at="Fluent English at a speed that keeps a rewrite under twenty seconds, on a "
                "laptop with 16 GB of memory. This is the model every number in the README "
                "was measured with.",
        watch_out="Its Italian is noticeably weaker than its English, and it slips on "
                  "grammar when asked to fill single words rather than rewrite a sentence.",
        languages="English well, other European languages adequately",
        measured="24 watermarked texts: detector z from 17.6 to 0.28, 74% of words changed, "
                 "5,800 tokens per 1,000 words, 17 seconds per text.",
        tags=["default", "fast"],
    ),
    Model(
        ref="qwen3:4b-instruct-2507-q8_0", params="4B", size_gb=4.3,
        good_at="The same model at a higher precision. Worth trying if the q4 version "
                "produces sentences that read slightly off.",
        watch_out="Nearly twice the download and noticeably slower on a machine that is "
                  "already short of memory.",
        languages="English well, other European languages adequately",
        tags=["quality"],
    ),
    Model(
        ref="gemma3:4b", params="4B", size_gb=3.3,
        good_at="Stronger on languages other than English than the 4B Qwen, which matters "
                "because the two Italian texts in the benchmark came out worst of all.",
        watch_out="Gemma is Google's, and Google watermarks the Gemini service. The open "
                  "weights do not carry it: SynthID is applied while sampling, and a local "
                  "server does not apply it unless you configure it to. Running the same "
                  "model through Google's API would mark the output.",
        languages="Strong multilingual",
        tags=["multilingual"],
    ),
    Model(
        ref="llama3.2:3b", params="3B", size_gb=2.0,
        good_at="The smallest download here that still rewrites whole paragraphs sensibly. "
                "A reasonable choice on a machine with 8 GB.",
        watch_out="Repeats phrasing from the original more than the Qwen models do, which "
                  "shows up directly as lower coverage.",
        languages="English, some European",
        tags=["small"],
    ),
    Model(
        ref="mistral-small:24b", params="24B", size_gb=14.0,
        good_at="The best prose of anything on this list, and good Italian and French.",
        watch_out="Fourteen gigabytes. On a 16 GB laptop it will page, and reflip will drop "
                  "to one request at a time. Sensible on 32 GB or more.",
        languages="Strong European languages",
        tags=["quality", "large"],
    ),
    Model(
        ref="qwen3:8b", params="8B", size_gb=5.2,
        good_at="Better sentences than the 4B when it finishes.",
        watch_out="It is a reasoning model: it thinks before it answers, and the thinking is "
                  "wasted on a paraphrase.",
        languages="English well, other European languages adequately",
        measured="Unusable on this laptop for this job: 464 seconds for one text, against "
                 "17 for the 4B instruct model.",
        tags=["slow"],
    ),
    Model(
        ref="hf.co/bartowski/Qwen2.5-7B-Instruct-GGUF:Q4_K_M", params="7B", size_gb=4.7,
        good_at="A middle size, pulled straight from Hugging Face rather than the Ollama "
                "library. Use it as the example of how to run anything that is published "
                "as a GGUF file.",
        watch_out="Not measured here. Any model you have not measured is a guess, and "
                  "`reflip models --measure` exists to stop it being one.",
        languages="English well, other European languages adequately",
        source="hf",
        tags=["hugging-face"],
    ),
]

# Never use these to rewrite: they mark their own output, so the result carries a
# watermark again, this time somebody else's.
REFUSED = {
    "claude": "Claude models launched since 2 August 2026 watermark their own text.",
    "gemini": "The Gemini app and API watermark their own text.",
    "anthropic": "Claude models launched since 2 August 2026 watermark their own text.",
}


def refusal(ref: str) -> str | None:
    """The sentence explaining why this model must not be the rewriter, if it must not."""
    low = ref.lower()
    for needle, why in REFUSED.items():
        if needle in low:
            return why
    return None


def recommended(installed: set[str] | None = None) -> list[dict]:
    """The catalogue, with each entry marked as downloaded or not."""
    installed = installed or set()
    out = []
    for m in CATALOGUE:
        d = m.to_dict()
        d["installed"] = m.ref in installed or m.ref.split(":")[0] in {i.split(":")[0] for i in installed}
        out.append(d)
    return out


def search(query: str, limit: int = 12) -> tuple[list[dict], str | None]:
    """Hugging Face repositories in GGUF form, most downloaded first.

    Returns (results, note). The note is a sentence when something is worth saying about
    the search rather than about a result, such as the network being unreachable.
    """
    params = urllib.parse.urlencode({
        "search": query, "filter": "gguf", "sort": "downloads",
        "direction": "-1", "limit": str(max(1, min(limit, 50))),
    })
    try:
        with urllib.request.urlopen(f"{HF_API}?{params}", timeout=HF_TIMEOUT) as r:
            found = json.loads(r.read().decode("utf-8"))
    except (urllib.error.URLError, OSError, ValueError) as e:
        return [], f"Hugging Face could not be reached: {' '.join(str(e).split())[:120]}"
    results = []
    for m in found:
        repo = m.get("id", "")
        results.append({
            "ref": f"hf.co/{repo}:Q4_K_M",
            "repo": repo,
            "downloads": m.get("downloads", 0),
            "likes": m.get("likes", 0),
            "gated": bool(m.get("gated")),
            "page": f"https://huggingface.co/{repo}",
            "refused": refusal(repo),
        })
    note = None
    if results:
        note = ("These are search results, not recommendations. Q4_K_M is assumed as the "
                "quantisation; check the repository for which files it actually has, and "
                "measure the model before trusting it.")
    return results, note


def quantisations(repo: str) -> tuple[list[str], str | None]:
    """The GGUF files a Hugging Face repository actually publishes."""
    try:
        with urllib.request.urlopen(f"{HF_API}/{repo}", timeout=HF_TIMEOUT) as r:
            data = json.loads(r.read().decode("utf-8"))
    except (urllib.error.URLError, OSError, ValueError) as e:
        return [], f"That repository could not be read: {' '.join(str(e).split())[:120]}"
    files = [f.get("rfilename", "") for f in data.get("siblings", [])]
    tags = sorted({f.rsplit("-", 1)[-1].removesuffix(".gguf") for f in files if f.endswith(".gguf")})
    if not tags:
        return [], "That repository has no GGUF files, so the local server cannot run it."
    return tags, None
